src/interceptor.ts

```diff
@@ -0,0 +1,91 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {CallOptions} from 'google-gax';
+import {OperationMetricsCollector} from './client-side-metrics/operation-metrics-collector';
+
+// Mock Server Implementation
+import * as grpcJs from '@grpc/grpc-js';
+import {status as GrpcStatus} from '@grpc/grpc-js';
+
+export type ServerStatus = {
+  metadata: {internalRepr: Map<string, Uint8Array[]>; options: {}};
+  code: number;
+  details: string;
+};
+
+// Helper to create interceptor provider for OperationMetricsCollector
+function createMetricsInterceptorProvider(
+  collector: OperationMetricsCollector,
+) {
+  return (options: grpcJs.InterceptorOptions, nextCall: grpcJs.NextCall) => {
+    let savedReceiveMessage: any;
+    // savedReceiveMetadata and savedReceiveStatus are not strictly needed here anymore for the interceptor's own state
+    // OperationStart and AttemptStart will be called by the calling code (`fakeReadModifyWriteRow`)
+    return new grpcJs.InterceptingCall(nextCall(options), {
+      start: (metadata, listener, next) => {
+        // AttemptStart is called by the orchestrating code
+        const newListener: grpcJs.Listener = {
+          onReceiveMetadata: (metadata, nextMd) => {
+            console.log('metadata encountered');
+            collector.onMetadataReceived(
+              metadata as unknown as {
+                internalRepr: Map<string, string[]>;
+                options: {};
+              },
+            );
+            nextMd(metadata);
+          },
+          onReceiveMessage: (message, nextMsg) => {
+            savedReceiveMessage = message; // Still need to know if a message was received for onResponse
+            nextMsg(message);
+          },
+          onReceiveStatus: (status, nextStat) => {
+            if (status.code === GrpcStatus.OK && savedReceiveMessage) {
+              collector.onResponse(); // Call onResponse for successful unary calls with a message
+            }
+            collector.onStatusMetadataReceived(
+              status as unknown as ServerStatus,
+            );
+            // AttemptComplete and OperationComplete will be called by the calling code
+            nextStat(status);
+          },
+        };
+        next(metadata, newListener);
+      },
+      sendMessage: (message, next) => next(message),
+      halfClose: next => next(),
+      cancel: next => next(),
+    });
+  };
+}
+
+export function withInterceptors(
+  gaxOptions: CallOptions,
+  metricsCollector?: OperationMetricsCollector,
+) {
+  if (metricsCollector) {
+    const interceptor = createMetricsInterceptorProvider(metricsCollector);
+    if (!gaxOptions.otherArgs) {
+      gaxOptions.otherArgs = {};
+    }
+    if (!gaxOptions.otherArgs.options) {
+      gaxOptions.otherArgs.options = {};
+    }
+    if (!gaxOptions.otherArgs.options.interceptors) {
+      gaxOptions.otherArgs.options.interceptors = [interceptor];
+    }
```

Comment:
don't you need to add the interceptor, even if other interceptors were added?

Reply:
That really depends on how we want to design the system. Here are the pros and cons of adding the interceptor.

Pros:
- Existing users that are already using interceptors, likely a small minority of our users, will automatically be using the new interceptor now so they don't have to add the CSM interceptor to their interceptors
Cons:
- It means the user doesn't have full control of the interceptors which reduces flexibility

Yeah, I guess the pros outweigh the cons so we can add this change.

Reply:
Actually, I'm not sure I like this change anymore. If the interceptors are not an array (although they should be) then trying to push another interceptor onto interceptors will throw an error. I don't want users to experience that just in case they are using interceptors incorrectly. As you can see in the screenshot, the compiler doesn't guarantee any particular structure for the interceptors property.

<img width="529" height="87" alt="image" src="https://github.com/user-attachments/assets/8236ad77-8ef6-4e56-804f-1ea8695aa3a9" />

Reply:
Just to make sure I'm understanding this right: with the current code, if the user passes in their own interceptors, they would effectively be disabling csm, right?

What else could they be passing other than an array? There must be some docstring somewhere telling us what we need to support

Reply:
All the [code examples](https://github.com/grpc/grpc-node/blob/01db2bc6206963a6678ee05fa4651fd2ee40068c/examples/interceptors/client.js#L108) pass interceptors in as an array. If you feel strongly about adding it to the existing interceptors then we can make that change, but I just don't like how the compiler doesn't guarantee that it is an array.

Reply:
Can we check the type of `gaxOptions.otherArgs.options.interceptors`, and only append if it is a list? And ignore unexpected input types?

I really don't think we should silently disable CSM if the user passes in an interceptor. That would present as a bug